messaging/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -335,33 +337,52 @@ public <Q, R> Stream<QueryResponseMessage<R>> scatterGather(@Nonnull QueryMessag
             return Stream.empty();
         }
 
-        long deadline = System.currentTimeMillis() + unit.toMillis(timeout);
-        return handlers.stream()
-                       .map(handler -> {
-                           Span handlerSpan = spanFactory.createInternalSpan(() -> "SimpleQueryBus.scatterGather(" + handlers.indexOf(handler) + ")", query)
-                                                         .start();
-                           long leftTimeout = getRemainingOfDeadline(deadline);
-                           ResultMessage<CompletableFuture<QueryResponseMessage<R>>> resultMessage =
-                                   interceptAndInvoke(DefaultUnitOfWork.startAndGet(interceptedQuery),
-                                                      handler);
-                           QueryResponseMessage<R> response = null;
-                           if (resultMessage.isExceptional()) {
-                               monitorCallback.reportFailure(resultMessage.exceptionResult());
-                               handlerSpan.recordException(resultMessage.exceptionResult());
-                               errorHandler.onError(resultMessage.exceptionResult(), interceptedQuery, handler);
-                           } else {
-                               try {
-                                   response = resultMessage.getPayload().get(leftTimeout, TimeUnit.MILLISECONDS);
-                                   monitorCallback.reportSuccess();
-                               } catch (Exception e) {
-                                   handlerSpan.recordException(e);
-                                   monitorCallback.reportFailure(e);
-                                   errorHandler.onError(e, interceptedQuery, handler);
-                               }
-                           }
-                           handlerSpan.end();
-                           return response;
-                       }).filter(Objects::nonNull);
+        return spanFactory.createInternalSpan(() -> "SimpleQueryBus.scatterGather", query).runSupplier(() -> {
```

Comment:
Curious: Why change the span layering here as you did?

Reply:
I tested this and it makes a lot more sense in the graph, to have one main scatterGather with various sub-gathers. 
However, the children span can be asynchronous in nature due to the stream, so they need to be created before to have the correct correlation

Reply:
Alright, thanks for explaining!